### PR TITLE
Expose 'limit' in RelationMeta and improve ‘scope’

### PR DESCRIPTION
**Pull request by mmanciop**
This PR adds support to `limit` in the `RelationMeta` class, which is currently not accessible at runtime by charms, and adds type validation and correctly sets the default value for `RelationMeta.scope`.

The [metadata.yaml spec](https://discourse.charmhub.io/t/charm-metadata-v2/3674) allows to limit the cardinality of relations using the `limit` keyword, e.g., when we want _up to_ one relation in the model using the `mongodb-datasource` interface. Additionally, the enumeration of the possible values for `RelationMeta.scope` (namely: `global`, which is default, or `container`) are not validated, and unless the scope is explicitly set, `RelationMeta.scope` is set to `None` rather than the default `global`.

### CODE REVIEW

File: test/test_charm.py [hunk 1]
```diff
@@ -299,6 +344,17 @@ def _get_action_test_meta(cls):
   description: "Start the unit."
 ''')
 
+    def test_relations_meta_limit_type_validation(self):
```
Comment (rbarry82) [inline test/test_charm.py]: Duplicate test.

File: ops/charm.py [hunk 2]
```diff
@@ -795,16 +795,35 @@ class RelationMeta:
         role: This is :class:`RelationRole`; one of peer/requires/provides
         relation_name: Name of this relation from metadata.yaml
         interface_name: Optional definition of the interface protocol.
-        scope: "global" or "container" scope based on how the relation should be used.
+        limit: Optional definition of maximum number of connections to this relation endpoint.
+        scope: "global" (default) or "container" scope based on how the relation should be used.
     """
 
+    VALID_SCOPES = ['global', 'container']
+    DEFAULT_SCOPE = VALID_SCOPES[0]
```
Comment (rbarry82) [inline ops/charm.py:803]: This is always dangerous. `DEFAULT_SCOPE` will be set a single time, at class evaluation time. It's not necessarily a big deal here, but every instance of the class will have the same values.

File: ops/charm.py [hunk 3]
```diff
@@ -795,16 +795,35 @@ class RelationMeta:
...
+                raise TypeError("limit should be an int, not {}".format(type(limit)))
+
+            self.limit = int(limit)
+        else:
+            self.limit = None
+
+        scope = raw.get('scope')
+        if scope is not None and scope not in RelationMeta.VALID_SCOPES:
+            valid_scopes = "'" + "', '".join(RelationMeta.VALID_SCOPES) + "'"
+            raise TypeError("scope should be one of {}; not '{}'".format(valid_scopes, scope))
```
Comment (rbarry82) [inline ops/charm.py:824]: This is confusing to read:

    raise TypeError("scope should be one of '{}'; not '{}'".format(
                    ', '.join(VALID_SCOPES), scope))

File: ops/charm.py [hunk 4]
```diff
@@ -795,16 +795,35 @@ class RelationMeta:
...
+
+        self.scope = scope or RelationMeta.DEFAULT_SCOPE
```
Comment (rbarry82) [inline ops/charm.py:826]: `self.scope = scope or DEFAULT_SCOPE`

File: ops/charm.py [hunk 5]
```diff
@@ -795,16 +795,35 @@ class RelationMeta:
...
         self.relation_name = relation_name
         self.interface_name = raw['interface']
-        self.scope = raw.get('scope')
+
+        limit = raw.get('limit')
+        if limit is not None:
+            if type(limit) is not int:
+                raise TypeError("limit should be an int, not {}".format(type(limit)))
+
+            self.limit = int(limit)
```
Comment (rbarry82) [inline ops/charm.py:817]: This will never be reached

File: ops/charm.py [hunk 6]
```diff
@@ -795,16 +795,35 @@ class RelationMeta:
...
     def __init__(self, role: RelationRole, relation_name: str, raw: dict):
         if not isinstance(role, RelationRole):
             raise TypeError("role should be a Role, not {!r}".format(role))
         self.role = role
         self.relation_name = relation_name
         self.interface_name = raw['interface']
-        self.scope = raw.get('scope')
+
+        limit = raw.get('limit')
+        if limit is not None:
```
Comment (rbarry82) [inline ops/charm.py:813]: `if limit and type(limit) is not int:`

Comment (rbarry82) [inline ops/charm.py:813]: `if limit and type(limit) is not int:`

**Review by rbarry82 — CHANGES_REQUESTED** [review]

**Review by rbarry82 — APPROVED** [review]
LGTM

File: ops/charm.py [hunk 7]
```diff
@@ -795,16 +795,28 @@ class RelationMeta:
...
         if not isinstance(role, RelationRole):
             raise TypeError("role should be a Role, not {!r}".format(role))
+        self._default_scope = self.VALID_SCOPES[0]
         self.role = role
         self.relation_name = relation_name
         self.interface_name = raw['interface']
-        self.scope = raw.get('scope')
+
+        self.limit = raw.get('limit')
+        if self.limit and type(self.limit) is not int:
```
Comment (jameinel) [inline ops/charm.py:813]: Isn't this better expressed as:
```
if self.limit and not isinstance(self.limit, int):
```

**Review by jameinel — APPROVED** [review]
Thanks. It does mean limit may be None or an integer, but I think that is well conveyed with the docstring "Optional ..."

And setting limit to 0 if it wasn't defined would be confusing, since "count > limit" would return true but 0 would actually mean infinite. So they already need a second if clause, and `if limit and count > limit` is a good way to express it.

**Review by balbirthomas — APPROVED** [review]

File: ops/charm.py [hunk 8]
```diff
@@ -795,16 +795,28 @@ class RelationMeta:
...
     def __init__(self, role: RelationRole, relation_name: str, raw: dict):
         if not isinstance(role, RelationRole):
             raise TypeError("role should be a Role, not {!r}".format(role))
+        self._default_scope = self.VALID_SCOPES[0]
         self.role = role
         self.relation_name = relation_name
         self.interface_name = raw['interface']
-        self.scope = raw.get('scope')
+
+        self.limit = raw.get('limit')
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply ops/charm.py:812]: Yeah I understand this to be the spec: https://discourse.charmhub.io/t/charm-metadata-v2/3674

File: ops/charm.py (hunk 7 above)
Comment (mmanciop) [reply ops/charm.py]: Done :-)

**Pull request merged**